### DISCUSSION TITLE
Transaction Schedule이란?

### DISCUSSION THREAD

**Discussion by J-I-H-O** [2024-woowacourse-study/level-interview#179] [Database]
.

**Comment by nayonsoso** [comment]
참고 : https://le2ksy.tistory.com/6

**Comment by le2sky** [comment]
트랜잭션 스케줄은 데이터베이스의 일관적인 상태를 유지하기 위해 동시에 실행되는 트랜잭션의 연산 순서를 정하는 것을 의미합니다.
트랜잭션의 연산 순서에 따라서 결과가 달라질 수 있기 때문에 트랜잭션을 병행 수행하기 위해서는 스케줄링이 중요합니다.
병행 수행에서 트랜잭션들은 차례대로 번갈아 가면서 수행되는 인터리빙 방식으로 진행됩니다.

트랜잭션 스케줄은 세 가지 유형이 존재합니다.

- 직렬 스케줄 :  인터리빙 방식을 이용하지 않고 트랜잭션별로 연산들을 순차 실행합니다. 격리성이 가장 높고, 처리 효율이 낮습니다. (실질적인 병행 수행이라고 보기 어려움)
- 비직렬 스케줄 : 인터리빙 방식을 이용해서 트랜잭션을 병행해서 수행합니다. 병행 수행으로 발생하는 갱신 분실, 모순성, 연쇄 복귀 문제가 발생할 수 있습니다.
- 직렬 가능 스케줄  :  직렬 스케줄에 따라 수행한 것과 같이 정확한 결과를 생성하는 비직렬 스케줄 방식입니다. 

트랜잭션 스케줄은 직렬 가능해야하며, 복구가 가능해야 트랜잭션을 병행 수행했을 때 발생하는 문제(갱신 분실, 모순성, 연쇄 복귀)를 예방할 수 있습니다.
트랜잭션을 병행수행하면서도 정확한 결과를 얻을 수 있는 직렬 가능성을 보장하기 위한 기법으로 DBMS는 병행 제어 기법을 사용합니다.
모든 트랜잭션이 준수하면 직렬 가능성이 보장되는 규약을 정의하고, 트랜잭션들이 해당 규약을 따르도록 합니다.
이러한 병행 제어 기법에는 대표적으로 로킹(locking) 기법이 있습니다.

**Comment by J-I-H-O** [comment]
트랜잭션 스케줄은 여러 트랜잭션이 병행 실행될 때, 각 트랜잭션의 연산들을 실행하는 순서를 정의한 것입니다.

**직렬 스케줄(Serial Schedule)**
    
![image](https://github.com/2024-woowacourse-study/level-interview/assets/110461155/5b41aba3-528b-4233-92ec-8ad5939680ef)
    
- 트랜잭션이 순차적으로 하나씩 실행됨
- 한 트랜잭션이 완료될 때 까지 다른 트랜잭션이 실행되지 않음
- 항상 일관성을 보장하지만, 병행 처리의 이점을 활용하지 못함

---

**비직렬 스케줄(Non-serial Schedule)**
    
![image](https://github.com/2024-woowacourse-study/level-interview/assets/110461155/da5d056c-a339-441b-9312-75d2c8a50202)
    
- 인터리빙 방식을 사용해 여러 트랜잭션이 병행하여 실행됨
    - 인터리빙(interleaving) : 트랜잭션을 연산 단위로 번걸아가며 수행하는 방식
- 병행 처리의 이점을 활용할 수 있지만, 적절히 관리되지 않으면 데이터베이스의 일관성이 깨질 수 있음
 
---

**직렬 가능 스케줄(Serializable Schedule)**
  - 직렬 스케줄과 동일한 결과를 산출하는 비직렬 스케줄
      - **서로 영향을 주지 않는 직렬 스케줄을 비직렬적으로 수행**하겠다는 것
      - 예시
          - 두 개의 트랜잭션이 Read 연산만을 수행하는 경우
          - 두 개의 트랜잭션이 같은 데이터 항목에 접근하지 않는 경우

비직렬 스케줄에서 데이터베이스의 일관성을 보장하기 위해 
트랜잭션들이 동일한 데이터 항목에 대해 임의적인 병행 접근을 하지 못하도록 제어하기 위한 로킹(Locking) 기법을 사용함
